gettsim/visualization.py

```diff
@@ -156,76 +109,158 @@ def plot_dag(
+
+    combo = pd.DataFrame(
+        {"x": node_x_coord, "y": node_y_coord, "url": url, "source_code": codes}
     )
+    combo.source_code = combo.source_code.str.split("\n").str.join("<br>")
 
-    graph_renderer.edge_renderer.visible = False
-    for (
-        _,
-        (start_node, end_node),
-    ) in graph_renderer.edge_renderer.data_source.to_df().iterrows():
-        (x_start, y_start), (x_end, y_end) = _compute_arrow_coordinates(
-            layout[start_node], layout[end_node]
+    # prepare for the edges dataframe
+    df = pd.DataFrame(list(dag.edges))
+    if len(df) == 0:
+        df["x0"] = 0
+        df["y0"] = 0
+    else:
+        df["x0"] = df[0].map(layout_df[0])
+        df["y0"] = df[0].map(layout_df[1])
+        df["x1"] = df[1].map(layout_df[0])
+        df["y1"] = df[1].map(layout_df[1])
+    df["None"] = ""
+    if len(df) == 0:
+        edge_x = []
+        edge_y = []
+    else:
+        edge_x = df[["x0", "x1", "None"]].apply(tuple, axis=1).tolist()
+        edge_x = list(reduce(operator.concat, edge_x))
+        edge_y = df[["y0", "y1", "None"]].apply(tuple, axis=1).tolist()
+        edge_y = list(reduce(operator.concat, edge_y))
+    # prepare for the arrows
+    arrows = []
+    for i in range(len(df)):
+        arrow = go.layout.Annotation(
+            x=df["x1"][i],
+            y=df["y1"][i],
+            xref="x",
+            yref="y",
+            text="",
+            showarrow=True,
+            axref="x",
+            ayref="y",
+            ax=df["x0"][i],
+            ay=df["y0"][i],
+            arrowhead=2,
+            arrowsize=2,
+            startstandoff=5,
+            standoff=5,
+            arrowcolor="gray",
         )
-        plot.add_layout(
-            Arrow(
-                end=NormalHead(**{**ARROW_KWARGS_DEFAULTS, **arrow_kwargs}),
-                x_start=x_start,
-                y_start=y_start,
-                x_end=x_end,
-                y_end=y_end,
-                **{**EDGE_KWARGS_DEFAULTS, **edge_kwargs},
-            )
+        arrows.append(arrow)
+
+    # plot the nodes, edges and arrows together
+
+    fig = go.FigureWidget(
+        layout=go.Layout(
+            showlegend=False,
+            hovermode="closest",
+            annotations=arrows,
+            hoverlabel_font_size=10,
+            margin={"b": 20, "l": 5, "r": 5, "t": 40},
+            xaxis={"showgrid": False, "zeroline": False, "showticklabels": False},
+            yaxis={"showgrid": False, "zeroline": False, "showticklabels": False},
         )
+    )
+    fig.add_scatter(
+        x=edge_x,
+        y=edge_y,
+        line={"width": 0.5, "color": "blue"},
+        hoverinfo="none",
+        mode="lines",
+        showlegend=False,
+    )
+    # choose the different option for plotting
+
+    if (len(names) > 10) or (show_labels is False):
+        mode = "markers"
+        hover_info = "text"
+    elif (len(names) <= 10) or ((len(names) > 10) and (show_labels is True)):
```

Comment:
Wouldn't `else` be enough here?

I like the simplification that within the if statements only the parameters are set!